File: src/android/org/sunbird/sync/model/HttpResponse.java
package org.sunbird.sync.model;

import android.text.TextUtils;
import android.util.Base64;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by swayangjit on 27/3/20.
 */
public class HttpResponse {
    private int status;
    private String url;
    private Map<String, List<String>> headers;
    private String body;
    private byte[] rawData;
    private JSONObject fileEntry;
    private boolean hasFailed;
    private boolean isFileOperation;
    private boolean isRawResponse;
    private String error;

    public void setStatus(int status) {
        this.status = status;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public void setHeaders(Map<String, List<String>> headers) {
        this.headers = headers;
    }

    public void setBody(String body) {
        this.body = body;
    }

    public void setData(byte[] rawData) {
        this.isRawResponse = true;
        this.rawData = rawData;
    }

    public void setFileEntry(JSONObject entry) {
        this.isFileOperation = true;
        this.fileEntry = entry;
    }

    public void setErrorMessage(String message) {
        this.hasFailed = true;
        this.error = message;
    }

    public boolean hasFailed() {
        return this.hasFailed;
    }

    public int getStatus() {
        return status;
    }

    public String getUrl() {
        return url;
    }

    public Map<String, List<String>> getHeaders() {
        return headers;
    }

    public String getBody() {
        return body;
    }

    public String getError() {
        return error;
    }

    public JSONObject toJSON() throws JSONException {
        JSONObject json = new JSONObject();

        json.put("status", this.status);
        json.put("url", this.url);

        if (this.headers != null && !this.headers.isEmpty()) {
            json.put("headers", new JSONObject(getFilteredHeaders()));
        }

        if (this.hasFailed) {
            json.put("error", this.error);
        } else if (this.isFileOperation) {
            json.put("file", this.fileEntry);
        } else if (this.isRawResponse) {
            json.put("data", Base64.encodeToString(this.rawData, Base64.DEFAULT));
        } else {
            json.put("data", this.body);
        }

        return json;
    }

    private Map<String, String> getFilteredHeaders() throws JSONException {
        Map<String, String> filteredHeaders = new HashMap<String, String>();

        for (Map.Entry<String, List<String>> entry : this.headers.entrySet()) {
            String key = entry.getKey();
            List<String> value = entry.getValue();

            if ((key != null) && (!value.isEmpty())) {
                filteredHeaders.put(key.toLowerCase(), TextUtils.join(", ", value));
            }
        }

        return filteredHeaders;
    }
}

File: src/ios/model/HttpResponse.swift
import Foundation

class HttpResponse {
    private var status: Int = 0
    private var url: String = ""
    private var headers: [String: [String]] = [:]
    private var body: String = ""
    private var rawData: [UInt8] = []
    private var fileEntry: [String: Any] = [:]
    private var hasFailed: Bool = false
    private var isFileOperation: Bool = false
    private var isRawResponse: Bool = false
    private var error: String = ""

    public func setStatus(_ status: Int) {
        self.status = status
    }

    public func setUrl(_ url: String) {
        self.url = url
    }

    public func setHeaders(_ headers: [String: [String]]) {
        self.headers = headers
    }

    public func setBody(_ body: String) {
        self.body = body
    }

    public func setData(_ rawData: [UInt8]) {
        self.isRawResponse = true
        self.rawData = rawData
    }

    public func setFileEntry(_ entry: [String: Any]) {
        self.isFileOperation = true
        self.fileEntry = entry
    }

    public func setErrorMessage(_ message: String) {
        self.hasFailed = true
        self.error = message
    }

    public func getHasFailed() -> Bool {
        return self.hasFailed
    }

    public func getStatus() -> Int {
        return self.status
    }

    public func  getUrl() -> String{
        return self.url
    }

    public func getHeaders() -> [String: String]{
        return self.headers
    }

    public func getBody() -> String {
        return self.body
    }

    public func getError() -> String {
        return self.error
    }

    public func toJSON() throws -> [String: Any] {
        var json: [String: Any] = [:]
        json["status"] = self.status
        json["url"] = self.url
        //TODO
//        if let headers = self.headers{
//            if !headers.isEmpty {
//                json["headers"] = getFilteredHeaders()
//            }
//        }
        if self.hasFailed {
            json["error"] = self.error
        } else if self.isFileOperation {
            json["file"] = self.fileEntry
        } else if self.isRawResponse {
            json["data"] = String(bytes: self.rawData, encoding: .utf8)
        } else {
            json["data"] = self.body
        }

        return json
    }

    private func getFilteredHeaders() -> [String: String] {
        var filteredHeaders: [String: String] = [:]
        // TODO Need to add the implementation 
        return filteredHeaders
    }
    
}
